keps/sig-instrumentation/1661-event-api-migration/README.md

```diff
@@ -0,0 +1,478 @@
+  // it acts on some changes in a ReplicaSet object.
+  // +optional
+  Regarding corev1.ObjectReference `json:"regarding,omitempty" protobuf:"bytes,8,opt,name=regarding"`
+
+  // Optional secondary object for more complex actions. E.g. when regarding object triggers
+  // a creation or deletion of related object.
+  // +optional
+  Related *corev1.ObjectReference `json:"related,omitempty" protobuf:"bytes,9,opt,name=related"`
+
+  // Optional. A human-readable description of the status of this operation.
+  // Maximal length of the note is 1kB, but libraries should be prepared to
+  // handle values up to 64kB.
+  // +optional
+  Note string `json:"note,omitempty" protobuf:"bytes,10,opt,name=note"`
+
+  // Type of this event (Normal, Warning), new types could be added in the
+  // future.
+  // +optional
+  Type string `json:"type,omitempty" protobuf:"bytes,11,opt,name=type"`
+
+  // Deprecated field assuring backward compatibility with core.v1 Event type
+  // +optional
+  DeprecatedSource corev1.EventSource `json:"deprecatedSource,omitempty" protobuf:"bytes,12,opt,name=deprecatedSource"`
+  // Deprecated field assuring backward compatibility with core.v1 Event type
+  // +optional
+  DeprecatedFirstTimestamp metav1.Time `json:"deprecatedFirstTimestamp,omitempty" protobuf:"bytes,13,opt,name=deprecatedFirstTimestamp"`
+  // Deprecated field assuring backward compatibility with core.v1 Event type
+  // +optional
+  DeprecatedLastTimestamp metav1.Time `json:"deprecatedLastTimestamp,omitempty" protobuf:"bytes,14,opt,name=deprecatedLastTimestamp"`
+  // Deprecated field assuring backward compatibility with core.v1 Event type
+  // +optional
+  DeprecatedCount int32 `json:"deprecatedCount,omitempty" protobuf:"varint,15,opt,name=deprecatedCount"`
+}
+
+// EventSeries contain information on series of events, i.e. thing that was/is happening
+// continuously for some time.
+type EventSeries struct {
+  // Number of occurrences in this series up to the last heartbeat time
+  Count int32 `json:"count" protobuf:"varint,1,opt,name=count"`
+  // Time when last Event from the series was seen before last heartbeat.
+  LastObservedTime metav1.MicroTime `json:"lastObservedTime" protobuf:"bytes,2,opt,name=lastObservedTime"`
+  // Information whether this series is ongoing or finished.
+  // Deprecated. Planned removal for 1.18
+  State EventSeriesState `json:"state" protobuf:"bytes,3,opt,name=state"`
+}
+
+type EventSeriesState string
+
+const (
+  EventSeriesStateOngoing  EventSeriesState = "Ongoing"
+  EventSeriesStateFinished EventSeriesState = "Finished"
+  EventSeriesStateUnknown  EventSeriesState = "Unknown"
+)
+```
+
+#### Few Examples
+
+| Regarding | Action | Reason | ReportingController | Related | 
+| ----------| -------| -------| --------------------|---------| 
+| Node X | BecameUnreachable | HeartbeatTooOld | kubernetes.io/node-ctrl | <nil> |
+| Node Y | FailedToAttachVolume | Unknown | kubernetes.io/pv-attach-ctrl | PVC X |
```

Comment:
| Regarding | Action  | Related |
| ----------| -------| ---------| 
| Node Y | FailedToAttachVolume  | PVC X |
| Pod X | ScheduledOn | Node Y |

The FailedToAttachVolume event is saying we failed to attach PVC to Node.  The ScheduledOn event is saying we scheduled Pod to Node.  They seem to be quite similar (put FOO on Node), so why are reversing Regarding vs Related?

Reply:
I don't think those examples really matter for this proposal (I would personally remove all of them, but given they are, let's leave them).

But to answer your question - you're probably right, it should be reversed. I think when Marek was writing that initially 3 years ago or sth, he thought that attaching is happening on the node, which isn't true - it's done in attach-detach controller. 
So yes - it should probably be reversed.

Reply:
Oh, wow.  This is old then...

Reply:
I looked at [the current state](https://github.com/kubernetes/kubernetes/blob/f52cbea102bb6ba9fe951cbf81e7dd4b48f95d9e/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L310) of the event, and it is already associated with Pod, so the discrepancy is just because the doc it is old.

I was trying to figure out if we were planning to change any event <-> object associations as part of the migration, and was also trying to see if the association rules are strict, or if they are closer to guidelines than hard rules.  "Tracing" is mentioned above, which means we would need strict rules for how we select the "Regarding" field, as that is generally how one would try and form a meaningful tree of events.  I also notice we mention making events more machine-readable an alternative below, so that makes it sound like we aren't trying to do anything like tracing...

Reply:
>  and was also trying to see if the association rules are strict, or if they are closer to guidelines than hard rules.

We're not going to have any strict rules. It's all the event`s author decision how to make (though they should carefully thing what related and regarding should really be - I expect that when we will be going over existing events, some adjustments will be needed wrt Related).

>  "Tracing" is mentioned above, which means we would need strict rules for how we select the "Regarding" field, as that is generally how one would try and form a meaningful tree of events. 

I don't think we should think about it as anything "tracing"-related. Events should give you ability to look into what was happening with your object, but tracing per-se is much bigger thing - we are not planning to do introduce any "why/reason" links between them. That should be left purely for the tracing effort.
I think think it's just a wording issue - we should change "tracing" to "tracking" or sth like that.

Reply:
Sounds good.  Thanks for the clarification.

Reply:
Thanks! I've reverse Related and Regarding for FailedToAttachVolume.